### PR TITLE
feat: alerts/reports add/edit modal

### PR DESCRIPTION
**Pull request by riahk**
### SUMMARY
- [x] Add `AlertReportModal`
- [x] Add components for ant-d `Select` and `Radio`

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Screen Shot 2020-12-07 at 2 18 32 PM
Screen Shot 2020-12-07 at 2 17 49 PM

### TEST PLAN
- [ ] Add spec file for `AlertReportModal`

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [x] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/config.py [hunk 1]
```diff
@@ -797,7 +797,7 @@ class CeleryConfig:  # pylint: disable=too-few-public-methods
 # Enable / disable Alerts, where users can define custom SQL that
 # will send emails with screenshots of charts or dashboards periodically
 # if it meets the criteria
-ENABLE_ALERTS = False
+ENABLE_ALERTS = True
```
Comment (nytai) [inline superset/config.py:800]: Do we want to switch this default?

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 2]
```diff
@@ -0,0 +1,1307 @@
...
+  onAdd?: (alert?: AlertObject) => void;
+  onHide: () => void;
+  show: boolean;
+}
+
+const NOTIFICATION_METHODS: NotificationMethod[] = ['Email', 'Slack'];
+
+const CONDITIONS = [
+  {
+    label: '< (Smaller than)',
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:53]: we should probably wrap these in `t()` for translation

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 3]
```diff
@@ -0,0 +1,1307 @@
...
+
+    data[name] = value;
+    setCurrentAlert(data);
+  };
+
+  // Handle input/textarea updates
+  const onTextChange = (
+    event:
+      | React.ChangeEvent<HTMLTextAreaElement>
+      | React.ChangeEvent<HTMLInputElement>,
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:763]: could also probably do `React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>` to be more terse

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 4]
```diff
@@ -0,0 +1,1307 @@
...
+              : undefined,
+            // log_retention: { value: resource.log_retention },
+            owners: (resource.owners || []).map(owner => ({
+              value: owner.id,
+              label: `${(owner as Owner).first_name} ${
+                (owner as Owner).last_name
+              }`,
+            })),
+            // @ts-ignore: Type not assignable
+            validator_config_json: JSON.parse(resource.validator_config_json),
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:923]: if you assign this to a var with a type, you should be able to assign it. Eg,

```typescript
const validatorConfig: AlertObject["validator_config_json"] = JSON.parse(resource.validator_config_json)

...

validator_config_json: validatorConfig
```

or `(JSON.parse(resource.validator_config_json) as AlertObject["validator_config_json"])`

File: superset/reports/commands/update.py [hunk 5]
```diff
@@ -87,6 +88,11 @@ def validate(self) -> None:
         # Validate chart or dashboard relations
         self.validate_chart_dashboard(exceptions, update=True)
 
+        if "validator_config_json" in self._properties:
+            self._properties["validator_config_json"] = json.dumps(
+                self._properties["validator_config_json"]
+            )
+
```
Comment (nytai) [inline superset/reports/commands/update.py:95]: wondering if we should handle this client side before we make the request. It is confusing having json inside json. Or perhaps add a check to see if `validator_config_json` is already a json string before running this, that way the api would support both (json string or json object).

**Comment by nytai** [comment]
Screen Shot 2020-12-08 at 5 56 27 PM

We should probably set a `min-width` and `overflow-x: auto` on the modal.

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx (hunk 4 above)
Comment (riahk) [reply superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:923]: I think the issue here if more that `resource` is expecting `validator_config_json` to be an object, not a string. I considered adding `string` to the type definition but it ended up creating more errors/checks I would need to handle, so ignoring this one case with the stringified response felt like less work in the long run.

File: superset/reports/commands/update.py (hunk 5 above)
Comment (riahk) [reply superset/reports/commands/update.py:95]: @dpgaspar added this logic to make it easier for the FE (you can see it in the `create` logic already), by not requiring us to stringify our json beforehand, which I'm fine with. Do you see it as being an issue in the long run?

Comment (nytai) [reply superset/reports/commands/update.py:95]: My main concern is with the api spec (ie, `/swagger/v1`). If it specifies a json string for this field a user will likely make an api request with a json string and possibly encounter a opaque error message from the api server.

Comment (dpgaspar) [reply superset/reports/commands/update.py:95]: I think that the OpenAPI spec is clear on this, it's just a nested structure, this happens on other places also. I think it's more elegant then sending a JSON string for a specific field

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx (hunk 4 above)
Comment (nytai) [reply superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:923]: right, I don't think having to always serialize the validator config object to json before adding it to an `AlertObject` is desirable or practical. This suggestion is to get rid of the `ts-ignore` comment by casting the output of the json parse to the desired type or to store it in typed var. 

https://stackoverflow.com/questions/38688822/how-to-parse-json-string-in-typescript

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx (hunk 2 above)
Comment (riahk) [reply superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:53]: opting to add the `t()` wrap in the Option component render below!

Comment (nytai) [reply superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:53]: Ah, that won't work. Translations only work with string literals that get extracted from source code into message catalogs which are then translated. The person translating the messages wouldn't know what they're translating if the message could be any string

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 6]
```diff
@@ -0,0 +1,1317 @@
...
+  show,
+  alert = null,
+  isReport = false,
+}) => {
+  const [disableSave, setDisableSave] = useState<boolean>(true);
+  const [currentAlert, setCurrentAlert] = useState<AlertObject | null>();
+  const [isHidden, setIsHidden] = useState<boolean>(true);
+  const [contentType, setContentType] = useState<string>(
+    alert && alert.chart ? 'chart' : 'dashboard',
+  ); // Maybe make this always default to dashboard and don't update until alert is fetched
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:452]: Do we want to do this? If not, let's remove the comment.

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 7]
```diff
@@ -0,0 +1,1317 @@
...
+  );
+
+  // Dropdown options
+  const [sourceOptions, setSourceOptions] = useState<MetaObject[]>([]);
+  const [dashboardOptions, setDashboardOptions] = useState<MetaObject[]>([]);
+  const [chartOptions, setChartOptions] = useState<MetaObject[]>([]);
+
+  const isEditMode = alert !== null;
+
+  // TODO: need to set status/settings list based on alert's notification settings
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:464]: Do we still need to do this?

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 8]
```diff
@@ -0,0 +1,1317 @@
...
+  const [notificationSettings, setNotificationSettings] = useState<
+    NotificationSetting[]
+  >([]);
+
+  const onNotificationAdd = () => {
+    const settings: NotificationSetting[] = notificationSettings.slice();
+
+    settings.push({
+      recipients: '',
+      options: NOTIFICATION_METHODS, // Need better logic for this
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:477]: Do we still need this comment?

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 9]
```diff
@@ -0,0 +1,1317 @@
...
+                ? getDashboardData(resource.dashboard) || {
+                    value: resource.dashboard.id,
+                  }
+                : undefined,
+              database: resource.database
+                ? getSourceData(resource.database) || {
+                    value: resource.database.id,
+                  }
+                : undefined,
+              // log_retention: { value: resource.log_retention },
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:917]: can we remove this? Looks like commented out code.

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 10]
```diff
@@ -0,0 +1,1317 @@
...
+            });
+          }
+        })
+        .catch(console.log);
+    }
+  } else if (
+    !isEditMode &&
+    (!currentAlert || currentAlert.id || (isHidden && show))
+  ) {
+    // TODO: update to match expected type variables
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:938]: Do we still need to do this?

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 11]
```diff
@@ -0,0 +1,1317 @@
...
+            <div className="input-container">
+              <AsyncSelect
+                name="owners"
+                isMulti
+                value={currentAlert ? currentAlert.owners : []}
+                loadOptions={loadOwnerOptions}
+                defaultOptions // load options on render
+                cacheOptions
+                onChange={onOwnersChange}
+                // disabled={!isDashboardLoaded}
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:1050]: can we remove this commented out code?

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 12]
```diff
@@ -0,0 +1,1317 @@
...
+                filterOption={null} // options are filtered at the api
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:1051]: This should still have the same behavior if you omit this prop

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 13]
```diff
@@ -0,0 +1,1317 @@
...
+                            value: currentAlert.database.value,
+                            label: currentAlert.database.label,
+                          }
+                        : undefined
+                    }
+                    loadOptions={loadSourceOptions}
+                    defaultOptions // load options on render
+                    cacheOptions
+                    onChange={onSourceChange}
+                    // disabled={!isDashboardLoaded}
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:1101]: ditto comment above

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 14]
```diff
@@ -0,0 +1,1317 @@
...
+                    filterOption={null} // options are filtered at the api
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:1102]: ditto comment above

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 15]
```diff
@@ -0,0 +1,1317 @@
...
+                      value: currentAlert.chart.value,
+                      label: currentAlert.chart.label,
+                    }
+                  : undefined
+              }
+              loadOptions={loadChartOptions}
+              defaultOptions // load options on render
+              cacheOptions
+              onChange={onChartChange}
+              filterOption={null} // options are filtered at the api
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:1268]: ditto comment above

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 16]
```diff
@@ -0,0 +1,1317 @@
...
+                      value: currentAlert.dashboard.value,
+                      label: currentAlert.dashboard.label,
+                    }
+                  : undefined
+              }
+              loadOptions={loadDashboardOptions}
+              defaultOptions // load options on render
+              cacheOptions
+              onChange={onDashboardChange}
+              filterOption={null} // options are filtered at the api
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:1289]: ditto comment above

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx [hunk 17]
```diff
@@ -0,0 +1,1317 @@
+/**
```
Comment (nytai) [inline superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:1]: this is a very large file, it's kind of hard to parse out the various pieces. Moving the styled components into a seperate file would help. Separating out some of the stateful logic from the jsx markup would also help. 

We're planning another round of changes for this feature so we can wait to address this then.

File: superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx (hunk 8 above)
Comment (riahk) [reply superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:477]: Kinda. I wanted to update the options list based on whether the user already has email or slack selected, but I think that can be added later.

Comment (nytai) [reply superset-frontend/src/views/CRUD/alert/AlertReportModal.tsx:477]: ok, let's swap out that ambiguous comment for a `TODO:` ☝️

**Review by nytai — APPROVED** [review]
LGTM
